### PR TITLE
Add auxiliary preprocessing regression gate

### LINKED ISSUE #10: Auxiliary preprocessor: stabilize and benchmark before broad trust
## Goal

Add a small internal regression and benchmark gate before relying broadly on the auxiliary-system preprocessor.

## Implementation

- Use synthetic Rust problems with known auxiliary systems.
- Use selected existing NL fixtures where equality-only auxiliary subsystems are present.
- Compare behavior with and without preprocessing:
  - status
  - objective
  - full constraint violation
  - iteration count
- Keep this as a regression/benchmark harness, not a public integration or decomposition API.
- Document only that `enable_preprocessing` may internally solve and remove auxiliary equality systems.
- Do not document decomposition details as a user-facing API.

## Acceptance

- no regressions on existing tests
- reduced path never returns a worse status than fallback for selected fixtures
- benchmark data is stable enough to guide whether the preprocessor should remain enabled by default

## Deferred validation anchors

After the Rust implementation is stable, optional validation scripts can compare graph partitions against:

- Pyomo incidence analysis: https://pyomo.readthedocs.io/en/stable/api/pyomo.contrib.incidence_analysis.html
- MathProgIncidence.jl: https://lanl-ansi.github.io/MathProgIncidence.jl/dev/overview.html

These are validation references only and must not become runtime dependencies.

### PR DESCRIPTION
**Pull request by bernalde**
Summary
- Adds an internal auxiliary preprocessing regression gate that compares preprocessed and no-preprocessing solves on synthetic Rust problems.
- Adds a small NL fixture gate that validates the preprocessed path against the fallback path on a parsed `.nl` problem.
- Expands structural incidence parity coverage translated from the local Pyomo, MathProgIncidence, and incidence_examples references, covering incidence adjacency/order, maximum matching, Dulmage-Mendelsohn partitions, connected components, and block triangular decomposition.
- Adds incidence_examples-specific fixtures for the tutorial singular-solid workflow, the CLC boundary-condition DM workflow, and the CLC SCC/topological block-initialization workflow.
- Keeps auxiliary decomposition details out of public-facing option/docs wording.

Tests run
- `PYTHONPATH=/home/bernalde/repos/incidence_examples MPLCONFIGDIR=/tmp/matplotlib-ripopt /home/bernalde/miniconda3/envs/lyopronto/bin/python -c "import idaes.core.util as u; from idaes.core.solvers import get_solver; u.get_solver = get_solver; import incidence_examples.tutorial.run_tutorial as t; ig=t.main(show=False, save=False); print(len(ig.constraints), len(ig.variables), len(ig.maximum_matching()))"` - passed; tutorial ended at `19 19 19`.
- `cargo test incidence_examples_tutorial_singular_solid_workflow -- --nocapture` - passed, 1 test.
- `cargo test incidence_examples_clc_dm_boundary_conditions_detect_wrong_endpoint_unfix -- --nocapture` - passed, 1 test.
- `cargo test incidence_examples_clc_scc_initialization_chain_blocks_are_ordered -- --nocapture` - passed, 1 test.
- `cargo test incidence_examples -- --nocapture` - passed, 4 tests.
- `rustfmt --check src/auxiliary_preprocessing.rs` - passed.
- `git diff --check` - passed.
- `cargo test` - passed, including 248 library tests, integration tests, and 1 doc test.
- `cargo nextest run` - passed, 367 tests run and 25 skipped.
- `make test-c` - passed, including release build and C API examples.

Notes
- The local `lyopronto` environment can run `incidence_examples` tutorial with an IDAES compatibility shim for `get_solver`.
- The larger incidence_examples CLC scripts cannot be executed directly in available local environments because the installed IDAES package does not provide `idaes.gas_solid_contactors`. Their structural behavior is represented as dependency-free Rust incidence fixtures instead.
- `gh pr view --web=false` currently fails with GitHub's classic Projects GraphQL deprecation error; `gh pr view --json ...` and `gh api` were used to inspect/update the PR instead.
- Local test output still shows pre-existing unreachable-code warnings in `tests/lbfgs_ipm.rs` and `examples/lbfgs_hessian.rs`, plus the existing nextest config warning.

Closes #10

### CODE REVIEW

**Comment by bernalde** [comment]
We definitely need to verify this implementation with https://pyomo.readthedocs.io/en/6.7.1/contributed_packages/incidence/tutorial.dm.html

**Comment by bernalde** [comment]
This repository contains 3 examples to verify this implementation https://github.com/Robbybp/incidence_examples

File: tests/ipm_paths.rs [hunk 1]
```diff
@@ -1107,6 +1107,204 @@ fn preprocessing_disabled_bypasses_auxiliary_preprocessing_in_solve() {
...
+    );
+    compare_auxiliary_gate_case(
+        "basin guard",
+        &AuxiliaryBasinGuardProblem,
+        &AuxiliaryBasinGuardProblem,
+    );
+}
+
+#[test]
+fn auxiliary_preprocessing_regression_gate_accepts_reduced_path_when_fallback_fails() {
```
Comment (bernalde) [inline tests/ipm_paths.rs:1305]: Nonblocking: This test can still pass if the preprocessed solve falls through to the same failing path as the no-preprocessing solve: `assert_auxiliary_gate_not_worse` treats equal `EvaluationError` statuses as acceptable and skips the objective/violation checks when neither side solved. That weakens this as a regression gate for the case named here. Please assert the intended asymmetry for this case, e.g. `is_solved(preprocessed.status)` and `!is_solved(fallback.status)`, or pass expected outcomes into `compare_auxiliary_gate_case`.

File: README.md [hunk 2]
```diff
@@ -460,8 +460,8 @@ Key options (all have Ipopt-matching defaults):
 | `warm_start`                    | false   | Enable warm-start initialization                           |
 | `constr_viol_tol`               | 1e-4    | Constraint violation tolerance                             |
 | `dual_inf_tol`                  | 1.0     | Dual infeasibility tolerance                               |
-| `enable_preprocessing`          | true    | Auxiliary equality blocks, fixed variables, redundancies   |
-| `auxiliary_tol`                 | 1e-8    | Accepted residual for auxiliary preprocessing block solves |
+| `enable_preprocessing`          | true    | Internal auxiliary equality solves, fixed variables, redundancies |
```
Comment (bernalde) [inline README.md:463]: Nonblocking: The README/Rust option docs now avoid the old auxiliary-block wording, but `ripopt --help` still prints `auxiliary_tol=<float> Auxiliary preprocessing residual tolerance` and `enable_preprocessing=<bool> Auxiliary blocks, fixed vars, redundancies` in `src/bin/ripopt_ampl.rs`. Since that help text is public-facing for AMPL users, please update those strings too so the CLI, README, and `SolverOptions` docs describe the option consistently.

**Review by bernalde — COMMENTED** [review]
1. Blocking issues: None.
2. Nonblocking issues: Two inline comments: strengthen the fallback-failure regression gate so it cannot pass when both paths fail, and align the AMPL CLI help wording with the updated README/Rust option docs. Also consider updating or de-emphasizing the README test-count section, which still says 230 tests while this branch now reports substantially more.
3. Questions: None.
4. Tests run and outcomes: targeted auxiliary gate tests passed; targeted NL fixture gate passed; `cargo test mathprogincidence -- --nocapture` passed; `cargo test` passed; `cargo nextest run` passed with 354 passed and 25 skipped; `cargo test --doc` passed; `cargo build --release` passed; `make test-c` passed. `cargo fmt --check` failed on existing repository-wide formatting drift, and an isolated rustfmt check also reports drift in touched test files.
5. Merge as-is: Yes. No blocking issues were found, and the nonblocking comments do not need to hold the merge.

Note: I attempted to submit this as APPROVE, but GitHub rejected it because the authenticated reviewer is the PR author, so this is submitted as COMMENT.

File: tests/ipm_paths.rs (hunk 1 above)
Comment (bernalde) [reply tests/ipm_paths.rs:1305]: Addressed in https://github.com/bernalde/ripopt/commit/c70dd29d8569ed43b04760715386c3912ee05166. The helper now returns both metrics, and `auxiliary_preprocessing_regression_gate_accepts_reduced_path_when_fallback_fails` explicitly asserts that preprocessing solves while the no-preprocessing path does not: https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/tests/ipm_paths.rs#L1305. Verified with `cargo test --test ipm_paths auxiliary_preprocessing_regression_gate -- --nocapture` and the broader local suite.

File: README.md (hunk 2 above)
Comment (bernalde) [reply README.md:463]: Addressed in https://github.com/bernalde/ripopt/commit/c70dd29d8569ed43b04760715386c3912ee05166. The AMPL help text now uses the same internal auxiliary equality solve wording as the README and SolverOptions docs for `auxiliary_tol` and `enable_preprocessing`: https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/src/bin/ripopt_ampl.rs#L143 and https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/README.md#L625. Verified with `cargo build --release`, `cargo test`, and CI.

**Comment by bernalde** [comment]
Replying to https://github.com/bernalde/ripopt/pull/21#issuecomment-4364810944: Addressed in https://github.com/bernalde/ripopt/commit/c70dd29d8569ed43b04760715386c3912ee05166. Added structural parity tests that mirror the Pyomo incidence tutorial DM chemical-looping cases: the singular model partition and the fixed square model with a valid block triangular decomposition. Relevant tests start at https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/src/auxiliary_preprocessing.rs#L3127. Verified with `cargo test pyomo_tutorial -- --nocapture` and the broader local/CI suites.

**Comment by bernalde** [comment]
Replying to https://github.com/bernalde/ripopt/pull/21#issuecomment-4364811818: Addressed at the structural incidence layer in https://github.com/bernalde/ripopt/commit/c70dd29d8569ed43b04760715386c3912ee05166. Added tests mirroring the lightweight incidence_examples matching gallery cases, and the Pyomo/tutorial tests above cover the DM and BTD behavior used by the repository examples: https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/src/auxiliary_preprocessing.rs#L3246. I did not vendor or run the IDAES example scripts directly because that would introduce heavy optional Python dependencies outside this Rust crate; the in-repo tests preserve the structural behavior this PR relies on. Verified with `cargo test incidence_examples -- --nocapture`, `cargo test mathprogincidence -- --nocapture`, and the broader local/CI suites.

**Comment by bernalde** [comment]
Replying to review summary https://github.com/bernalde/ripopt/pull/21#pullrequestreview-4215880610: Addressed in https://github.com/bernalde/ripopt/commit/c70dd29d8569ed43b04760715386c3912ee05166. The fallback-failure gate and AMPL help wording were updated, and the README tests/coverage section no longer carries stale exact test counts: https://github.com/bernalde/ripopt/blob/c70dd29d8569ed43b04760715386c3912ee05166/README.md#L771. Formatting note: `rustfmt --check src/auxiliary_preprocessing.rs` and `git diff --check` pass; I did not include a whole-file formatting sweep for unrelated existing drift. Local tests and the GitHub `test` check passed.
